### DISCUSSION TITLE
Wrapping this in simple use class

### DISCUSSION THREAD

**Discussion by dwhiteddsoft** [bytebeamio/rumqtt#280] [Q&A]
Is there any inherent problems getting the client and event loop into a class? I am kinda new to Rust but was simply trying to wrap all this up in a struct impl so I could hide it away from the main. Something simple like my_sruct::start

But for what ever reason when I move the same code works in the mainloop to a struct impl I can never even get connected? Any thoughts? I know this is vague and I am ok with vague answers or asks for clarity

Thanks in advance

**Comment by de-sh** [comment]
Hey @dwhiteddsoft, we are working on implementing an abstraction that would work something like the following:
Subscribing:
```rust
let subscriber: Subscriber = client.subscriber("topic/string", QoS);
while let Some(pub) = subscriber.next().await { ... }
```
Publishing:
```rust
let publisher: Publisher = client.publisher("topic/string", QoS);
publisher.publish(data, retained).await;
```

I believe this is something similar to what you are suggesting above, so what are your thoughts on this?

**Reply by de-sh** [reply]
Here's a PoC in code: https://github.com/bytebeamio/rumqtt/commit/150cec51553ef73c6e5128a1ba39df3141bf7792#diff-fd67087e30ee8bbba41e6c314cab2aea694c64ea2cd37b27cff20fb2f2f1a650

**Reply by dwhiteddsoft** [reply]
I like what you wrote in the comments section here. I especially like the Subscriber class. I am a little confused by the new example you linked though as I do not see the clean Subscriber demonstrated. I love the work though. Thanks!

**Reply by de-sh** [reply]
Understood, have added this and a few other features that will make working with the abstractions a bit easier, let us know what you think: 
- [Publisher](https://github.com/bytebeamio/rumqtt/commit/b25ba087792851df04f7d8c01e414408d001b406#diff-4af5ded3c03e4a350811d42586b47732f0096e95f97d56368ea80c9f8da47833)
- [Subscriber](https://github.com/bytebeamio/rumqtt/commit/b25ba087792851df04f7d8c01e414408d001b406#diff-484e877da7d3dae8d95d0a28b29e2ac144edd224cd617727fc5c56ef3889d719)
- [Unsubscribe per subscriber](https://github.com/bytebeamio/rumqtt/commit/b25ba087792851df04f7d8c01e414408d001b406#diff-ea7127d2299baa717e8894da7fc3523c86f63b74f1b06a8af37ca45ba17e2f85R125)
- [Disconnect on drop](https://github.com/bytebeamio/rumqtt/commit/b25ba087792851df04f7d8c01e414408d001b406#diff-ea7127d2299baa717e8894da7fc3523c86f63b74f1b06a8af37ca45ba17e2f85R132)
